src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1057,958 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+            original_stderr = sys.stderr
+            f = open(os.devnull, 'w')
+            sys.stderr = f
+
+            make_api_call_with_retries(identity_uri, identity_data, False, consts.Post_RefreshToken_Fault_Type,
+                                       'Unable to post refresh token details to clientproxy',
+                                       "Failed to pass refresh token details to proxy.", clientproxy_process)
+            sys.stderr = original_stderr
+
+    data = prepare_clientproxy_data(response)
+    expiry = data['hybridConnectionConfig']['expirationTime']
+
+    if token is not None:
+        data['kubeconfigs'][0]['value'] = insert_token_in_kubeconfig(data, token)
+
+    uri = f'http://localhost:{client_proxy_port}/subscriptions/{subscription_id}/resourceGroups/{resource_group_name}/providers/Microsoft.Kubernetes/connectedClusters/{cluster_name}/register?api-version=2020-10-01'
+
+    # Posting hybrid connection details to proxy in order to get kubeconfig
+    response = make_api_call_with_retries(uri, data, False, consts.Post_Hybridconn_Fault_Type,
+                                          'Unable to post hybrid connection details to clientproxy',
+                                          "Failed to pass hybrid connection details to proxy.", clientproxy_process)
+
+    if flag == 0:
+        # Decoding kubeconfig into a string
+        try:
+            kubeconfig = json.loads(response.text)
+        except Exception as e:
+            telemetry.set_user_fault()
+            telemetry.set_exception(exception=e, fault_type=consts.Load_Kubeconfig_Fault_Type,
+                                    summary='Unable to load Kubeconfig')
+            close_subprocess_and_raise_cli_error(clientproxy_process,"Failed to load kubeconfig." + str(e))
+        
+        kubeconfig = kubeconfig['kubeconfigs'][0]['value']
+        kubeconfig = b64decode(kubeconfig).decode("utf-8")
+
+        try:
+            print_or_merge_credentials(path, kubeconfig, True, context_name)
+            if context_name is None:
+                kubeconfig_obj = load_kubernetes_configuration(path)
+                temp_context_name = kubeconfig_obj['current-context']
+            else:
+                temp_context_name = context_name
+            print("You can now start sending requests using kubectl on '{}' context using kubeconfig at {}".format(temp_context_name, path))
+            print("Press Ctrl+C to close proxy.")
+
+        except Exception as e:
+            telemetry.set_exception(exception=e, fault_type=consts.Merge_Kubeconfig_Fault_Type,
+                                    summary='Unable to merge kubeconfig.')
+            close_subprocess_and_raise_cli_error(clientproxy_process, "Failed to merge kubeconfig." + str(e))
+    
+    return expiry, clientproxy_process
+
+def make_api_call_with_retries(uri, data, tls_verify, fault_type, summary, cli_error, clientproxy_process):
+    for i in range(consts.API_CALL_RETRIES):
+        try:
```

Comment:
Just add one sec sleep here.

Reply:
this is intentional. On linux, csp takes variable number of seconds for transitioning into listening state from running state. Hence the retries.